gossip/src/crds_value.rs

```diff
@@ -490,84 +496,193 @@ impl Sanitize for NodeInstance {
+            .map(|i| bits.get(i))
+            .dedup_with_count()
+            .map(|(count, _)| count)
+            .take_while(|count| Self::valid_and_has_space(&mut total_serde_bytes, *count))
+            .map(|count| U16(count.try_into().unwrap()))
+            .collect();
+        Self { encoded }
+    }
+
+    pub fn slots_count(&self) -> usize {
+        self.encoded.iter().map(|x| x.0 as usize).sum::<usize>()
+    }
+
+    pub fn to_slots(&self, last_slot: Slot, min_slot: Slot) -> Vec<Slot> {
+        let mut result: Vec<Slot> = self
+            .encoded
+            .iter()
+            .zip([1, 0].iter().cycle())
+            .flat_map(|(bit_count, bit)| std::iter::repeat(bit).take(bit_count.0.into()))
+            .enumerate()
+            .filter_map(|(offset, bit): (usize, &i32)| {
+                if *bit > 0 {
+                    Some(last_slot.saturating_sub(offset as Slot))
+                } else {
+                    None
+                }
+            })
+            .take(MAX_SLOTS_PER_ENTRY)
+            .take_while(|slot| *slot >= min_slot)
+            .collect();
+        result.reverse();
+        result
+    }
+}
+
+#[derive(Deserialize, Serialize, Clone, Debug, PartialEq, Eq, AbiExample)]
+pub struct RawOffsets {
+    bits: BitVec<u8>,
+}
+
+impl RawOffsets {
+    pub fn new(mut bits: BitVec<u8>) -> Self {
+        bits.truncate(RestartLastVotedForkSlots::MAX_SPACE as u64 * 8);
+        bits.shrink_to_fit();
+        Self { bits }
+    }
+
+    pub fn to_slots(&self, last_slot: Slot, min_slot: Slot) -> Vec<Slot> {
+        let mut result = Vec::new();
+        for i in 0..self.bits.len() {
+            if self.bits.get(i) {
+                let slot = last_slot.saturating_sub(i as Slot);
+                if slot < min_slot {
+                    return result;
+                }
+                result.push(slot);
+                if result.len() > MAX_SLOTS_PER_ENTRY {
+                    break;
+                }
+            }
+        }
+        result.reverse();
+        result
+    }
+}
+
+#[derive(Serialize, Deserialize, Clone, PartialEq, Eq, AbiExample, Debug)]
 pub struct RestartLastVotedForkSlots {
     pub from: Pubkey,
     pub wallclock: u64,
-    pub slots: Vec<CompressedSlots>,
+    offsets: SlotsOffsets,
+    pub last_voted_slot: Slot,
     pub last_voted_hash: Hash,
     pub shred_version: u16,
 }
 
 impl Sanitize for RestartLastVotedForkSlots {
     fn sanitize(&self) -> std::result::Result<(), SanitizeError> {
-        if self.slots.is_empty() {
-            return Err(SanitizeError::InvalidValue);
-        }
-        self.slots.sanitize()?;
         self.last_voted_hash.sanitize()
     }
 }
 
 impl RestartLastVotedForkSlots {
-    pub fn new(from: Pubkey, now: u64, last_voted_hash: Hash, shred_version: u16) -> Self {
-        Self {
+    // This number is MAX_CRDS_OBJECT_SIZE - empty serialized RestartLastVotedForkSlots.
+    const MAX_SPACE: usize = 824;
```

Comment:
Because above you are working with bits, can you rename this to `MAX_BYTES` to be clear this is "bytes"?

also is there a test that fills out this limit, and confirms that the crds object size stays below `MAX_CRDS_OBJECT_SIZE`?

Reply:
Done. Yes there is a test called test_restart_last_voted_fork_slots_max_bytes